### PR TITLE
Ports ability to upload honk virus to airlocks

### PR DESCRIPTION
**Pull request by Archanial**
## About The Pull Request

funny I liked it and signals are used in wiremod

ported pr:
https://github.com/tgstation/tgstation/pull/52593

## Changelog  
:cl: KathyRyals 
add: The Honkworks 5.0 cartridge's software has been updated to be able to hack machines, computers and airlocks to make them honk for a while.
code: Added a new sound_player component to be able to play any sound with any signals.
/:cl:

### CODE REVIEW

**Comment by KittyNoodle** [comment]
tfw no honk emote on github

File: code/game/objects/items/devices/PDA/PDA.dm [hunk 1]
```diff
@@ -903,6 +903,29 @@ GLOBAL_LIST_EMPTY(PDAs)
 		playsound(src, 'sound/machines/pda_button1.ogg', 50, TRUE)
 	return TRUE
 
+/obj/item/pda/pre_attack(obj/target, mob/living/user, params)
+	if(!ismachinery(target))
+		return ..()
+	var/obj/machinery/target_machine = target
+	if(!target_machine.panel_open && !istype(target, /obj/machinery/computer))
```
Comment (ike709) [inline code/game/objects/items/devices/PDA/PDA.dm:910]:
```suggestion
	var/obj/machinery/computer/target_machine = target
	if(!istype(target_machine) || !target_machine.panel_open)
```

File: code/game/objects/items/devices/PDA/PDA.dm [hunk 2]
```diff
@@ -903,6 +903,29 @@ GLOBAL_LIST_EMPTY(PDAs)
...
+		return ..()
+	if(!istype(cartridge, /obj/item/cartridge/virus/clown))
+		return ..()
+	var/obj/item/cartridge/virus/installed_cartridge = cartridge
+
+	if(installed_cartridge.charges <=0)
+		to_chat(user, "<span class='notice'>Out of charges.</span>")
+		return ..()
+	to_chat(user, "<span class='notice'>You upload the virus to the airlock controller!</span>")
+	var/sig_list
```
Comment (ike709) [inline code/game/objects/items/devices/PDA/PDA.dm]:
```suggestion
	var/list/sig_list = list()
```

File: code/game/objects/items/devices/PDA/PDA.dm [hunk 3]
```diff
@@ -903,6 +903,29 @@ GLOBAL_LIST_EMPTY(PDAs)
...
+	if(istype(target,/obj/machinery/door/airlock))
+		sig_list += list(COMSIG_AIRLOCK_OPEN, COMSIG_AIRLOCK_CLOSE)
+	else
+		sig_list += list(COMSIG_ATOM_ATTACK_HAND)
+	target.AddComponent(/datum/component/sound_player, amount = (rand(15,20)), signal_or_sig_list = sig_list)
+	installed_cartridge.charges --
```
Comment (ike709) [inline code/game/objects/items/devices/PDA/PDA.dm]:
```suggestion
	installed_cartridge.charges--
```

File: code/datums/components/soundplayer.dm [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+/datum/component/sound_player/proc/play_sound()
+	playsound(parent, pickweight(sounds), volume, TRUE)
+	switch(amount_left)
+		if(-1)
+			return
+		if(1) //Last use.
+			qdel(src)
+			return
+		else
+			amount_left --
```
Comment (ike709) [inline code/datums/components/soundplayer.dm]:
```suggestion
			amount_left--
```

File: code/game/objects/items/devices/PDA/PDA.dm (hunk 1 above)
Comment (KubeRoot) [reply code/game/objects/items/devices/PDA/PDA.dm:910]: It looks like the original intent of the code is to succeed if either
- Target is a computer
- Target is a non-computer machine with the panel open

While this change makes it so the code succeeds only if **both**
- Target is a computer
- Target has panel open

Which I don't even know if is possible with computers?

File: code/game/objects/items/devices/PDA/PDA.dm [hunk 5]
```diff
@@ -903,6 +903,29 @@ GLOBAL_LIST_EMPTY(PDAs)
...
+	if(!target_machine.panel_open && !istype(target, /obj/machinery/computer))
+		return ..()
+	if(!istype(cartridge, /obj/item/cartridge/virus/clown))
+		return ..()
+	var/obj/item/cartridge/virus/installed_cartridge = cartridge
+
+	if(installed_cartridge.charges <=0)
+		to_chat(user, "<span class='notice'>Out of charges.</span>")
+		return ..()
+	to_chat(user, "<span class='notice'>You upload the virus to the airlock controller!</span>")
```
Comment (KubeRoot) [inline code/game/objects/items/devices/PDA/PDA.dm]: Is this message going to be displayed as-is even when hacking things other than airlocks?

File: code/game/objects/items/devices/PDA/PDA.dm (hunk 1 above)
Comment (ike709) [reply code/game/objects/items/devices/PDA/PDA.dm:910]: Ah I see.

File: code/game/objects/items/devices/PDA/PDA.dm (hunk 5 above)
Comment (Archanial) [reply code/game/objects/items/devices/PDA/PDA.dm]: For now it is limited to cartridge/virus/clown so only airlocks will display it

Comment (KubeRoot) [reply code/game/objects/items/devices/PDA/PDA.dm]: Not the case in the current state of the PR as far as I can tell.

image

Comment (Archanial) [reply code/game/objects/items/devices/PDA/PDA.dm]: That's what I get for copypasting tg lmao. Will fix right away.

Comment (Archanial) [reply code/game/objects/items/devices/PDA/PDA.dm]: Alright with next commit this shouldn't issue any longer.

**Review by ike709 — DISMISSED** [review]

File: code/game/objects/items/devices/PDA/PDA.dm [hunk 6]
```diff
@@ -903,6 +903,29 @@ GLOBAL_LIST_EMPTY(PDAs)
...
+		balloon_alert(user, "Out of charges")
+		return ..()
+	balloon_alert(user, "Virus uploaded")
+	var/list/sig_list = list()
+	if(istype(target,/obj/machinery/door/airlock))
+		sig_list += list(COMSIG_AIRLOCK_OPEN, COMSIG_AIRLOCK_CLOSE)
+	else
+		sig_list += list(COMSIG_ATOM_ATTACK_HAND)
+	installed_cartridge.charges--
+	target.AddComponent(/datum/component/sound_player, amount = (rand(30,50)), signal_or_sig_list = sig_list)
```
Comment (PowerfulBacon) [inline code/game/objects/items/devices/PDA/PDA.dm:931]: What if it already has the sound player component?

**Review by PowerfulBacon — DISMISSED** [review]
Seems fine but idk how this works if you hack an already hacked airlock

File: code/game/objects/items/devices/PDA/PDA.dm [hunk 7]
```diff
@@ -903,6 +903,29 @@ GLOBAL_LIST_EMPTY(PDAs)
...
+	if(!istype(cartridge, /obj/item/cartridge/virus/clown))
+		return ..()
+	var/obj/item/cartridge/virus/installed_cartridge = cartridge
+
+	if(installed_cartridge.charges <=0)
+		balloon_alert(user, "Out of charges")
+		return ..()
+	balloon_alert(user, "Virus uploaded")
+	var/list/sig_list = list()
+	if(istype(target,/obj/machinery/door/airlock))
```
Comment (Archanial) [inline code/game/objects/items/devices/PDA/PDA.dm]:
```suggestion
	if(istype(target, /obj/machinery/door/airlock))
```

**Review by ike709 — DISMISSED** [review]

File: code/game/objects/items/devices/PDA/PDA.dm [hunk 8]
```diff
@@ -916,6 +916,10 @@ GLOBAL_LIST_EMPTY(PDAs)
 	if(installed_cartridge.charges <=0)
 		balloon_alert(user, "Out of charges")
 		return ..()
+
+	if(target.GetComponent(/datum/component/sound_player))
+		return
```
Comment (ike709) [inline code/game/objects/items/devices/PDA/PDA.dm]: Actually I feel like this should have feedback along the lines of "This is already hacked"

Comment (Archanial) [reply code/game/objects/items/devices/PDA/PDA.dm]:
```suggestion
	if(target.GetComponent(/datum/component/sound_player))
		balloon_alert(user, "This is already hacked")
		return
```

**Review by ike709 — APPROVED** [review]

**Review by PowerfulBacon — APPROVED** [review]

**Pull request merged**